crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,732 @@
+    }
+
+    fn fix_title(&self) -> String {
+        "Remove extraneous blank line(s)".to_string()
+    }
+}
+
+/// ## What it does
+/// Checks for missing blank lines after end of function or class.
+///
+/// ## Why is this bad?
+/// PEP 8 recommends the using blank lines as following:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// class User(object):
+///     pass
+/// user = User()
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class User(object):
+///     pass
+///
+///
+/// user = User()
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E305.html)
+#[violation]
+pub struct BlankLinesAfterFunctionOrClass(pub u32);
+
+impl AlwaysFixableViolation for BlankLinesAfterFunctionOrClass {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesAfterFunctionOrClass(blank_lines) = self;
+        format!("expected 2 blank lines after class or function definition, found ({blank_lines})")
+    }
+
+    fn fix_title(&self) -> String {
+        "Add missing blank line(s)".to_string()
+    }
+}
+
+/// ## What it does
+/// Checks for for 1 blank line between nested functions/classes definitions.
+///
+/// ## Why is this bad?
+/// PEP 8 recommends the using blank lines as following:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// def outer():
+///     def inner():
+///         pass
+///     def inner2():
+///         pass
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def outer():
+///     def inner():
+///         pass
+///
+///     def inner2():
+///         pass
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E306.html)
+#[violation]
+pub struct BlankLinesBeforeNestedDefinition(pub u32);
+
+impl AlwaysFixableViolation for BlankLinesBeforeNestedDefinition {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesBeforeNestedDefinition(blank_lines) = self;
+        format!("Expected 1 blank line before a nested definition, found {blank_lines}")
+    }
+
+    fn fix_title(&self) -> String {
+        "Add missing blank line".to_string()
+    }
+}
+
+/// Returns `true` if the token is Async, Class or Def
+fn is_top_level_token(token: &Option<Tok>) -> bool {
+    matches!(&token, Some(Tok::Class | Tok::Def))
+}
+
+/// Returns `true` if the token is At, Async, Class or Def
+fn is_top_level_token_or_decorator(token: &Tok) -> bool {
+    matches!(&token, Tok::Class | Tok::Def | Tok::At)
+}
+
+#[derive(Debug)]
+struct LogicalLineInfo {
+    first_token: Tok,
+    first_token_range: TextRange,
+    last_token: Tok,
+    last_token_range: TextRange,
```

Comment:
It seems to be sufficient to only store the end offset of the last token.
```suggestion
    last_token_end: TextSize,
```

Reply:
Done in 0f659b965.